source/simulator/nullspace.cc

```diff
@@ -81,135 +90,184 @@ namespace aspect
+    //the velocity correction if it is not selected by the NullspaceRemoval flag
+    if (use_constant_density) //disable translation correction
       {
-        Assert(introspection.block_indices.velocities != introspection.block_indices.pressure,
-               ExcNotImplemented());
-
-        for (unsigned int i=0; i<net_rotations_translations.size(); ++i)
-          {
-            // compute the magnitude of the solution vector in direction
-            // of this null space vector and subtract the corresponding multiple
-            const double power = net_rotations_translations[i]
-                                 * tmp_distributed_stokes.block(introspection.block_indices.velocities);
-            tmp_distributed_stokes.block(introspection.block_indices.velocities).sadd(1.0,
-                                                                                      -1.0*power,
-                                                                                      net_rotations_translations[i]);
-          }
-        relevant_dst.block(0) = tmp_distributed_stokes.block(0);
+        if( !(parameters.nullspace_removal & NullspaceRemoval::net_translation_x) )
+          velocity_correction[0] = 0.0;  //don't correct x translation
+        if( !(parameters.nullspace_removal & NullspaceRemoval::net_translation_y) )
+          velocity_correction[1] = 0.0;  //don't correct y translation
+        if( !(parameters.nullspace_removal & NullspaceRemoval::net_translation_z) && dim == 3 )
+          velocity_correction[2] = 0.0;  //don't correct z translation
       }
-    if (parameters.nullspace_removal & NullspaceRemoval::angular_momentum)
+    else //disable momentum correction
       {
-        remove_net_angular_momentum( relevant_dst, tmp_distributed_stokes);
+        if( !(parameters.nullspace_removal & NullspaceRemoval::linear_momentum_x) )
+          velocity_correction[0] = 0.0;  //don't correct x translation
+        if( !(parameters.nullspace_removal & NullspaceRemoval::linear_momentum_y) )
+          velocity_correction[1] = 0.0;  //don't correct y translation
+        if( !(parameters.nullspace_removal & NullspaceRemoval::linear_momentum_z) && dim == 3 )
+          velocity_correction[2] = 0.0;  //don't correct z translation
       }
-  }
 
+    // Now construct a translation vector with the desired rate and subtract it from our vector
+    internal::Translation<dim> translation( velocity_correction );
+    interpolate_onto_velocity_system(translation, correction);
+    tmp_distributed_stokes.block(introspection.block_indices.velocities).sadd(1.0, -1.0,correction);
 
-
-  template <>
-  void
-  Simulator<3>::remove_net_angular_momentum( LinearAlgebra::BlockVector &relevant_dst,
-                                             LinearAlgebra::BlockVector &tmp_distributed_stokes )
-  {
-    AssertThrow(false, ExcNotImplemented());
+    //copy into the locally relevant vector
+    relevant_dst.block(0) = tmp_distributed_stokes.block(0);
   }
```

Comment:
Last comment for today: This function looks quite reasonable to me. Thanks for cleaning it up -- I think it's much better than it was before! Will continue with the angular momentum today.